<center>
    <h1>
        <bold>
            Penerapan Algoritma Greedy untuk Menentukan Rute Terpendek dari Rumah Acep (Lembang) ke Kampus UNIKOM Dipatiukur
        </bold>
    </h1>
</center>

### Menampilkan NIM, NAMA, dan Kelas

---

| NIM | Nama | Kelas |
| -------- | -------- | -------- |
| 10121295 | Rifai Nugroho | IF8 |
| 10121307 | Cecep Ramdan Suryadi | IF8 |
| 10121311 | Rafly Sidiq Hermawan | IF8 |
| 10121314 | Acep Effendi Abdurrochman | IF8 |
| 10121318 | Bagus Aimar Atqansyah | IF8 |
| 10121907 | Rizki Ananda Saputra | IF8 |

---

### Menampilkan 2 Titik Koordinat  A dan L

In [8]:
import folium

# koordinat awal
TitikAwal = (-6.81458,107.62306) # Titik Koordinat Rumah Acep (Lembang)

# koordinat akhir
TitikAkhir = (-6.88701,107.61552) # Titik Koordinat Kampus UNIKOM (Dipatiukur)

# buat objek peta
peta = folium.Map(location=TitikAwal, zoom_start=12)

# menampilkan marker titik awal dan titik akhir dengan keterangan Node A dan Node L
folium.Marker(location=TitikAwal, tooltip="Node A", icon=folium.Icon(color='green')).add_to(peta)
folium.Marker(location=TitikAkhir, tooltip="Node L", icon=folium.Icon(color='red')).add_to(peta)

# tampilkan peta
peta

### Menampilkan Keterangan Sesuai Map di Atas 
---

| Node | Nama Jalan |
| -------- | -------- |
| A | Jalan Maribaya|
| L | Jalan Dipatiukur (UNIKOM)|

---

### Menampilkan 2 Titik Koordinat Lalu Hubungkan dengan Garis

In [9]:
import folium

# koordinat awal
TitikAwal = (-6.81458,107.62306) # Titik Koordinat Rumah Acep (Lembang)

# koordinat akhir
TitikAkhir = (-6.88701,107.61552) # Titik Koordinat Kampus UNIKOM (Dipatiukur)

# buat objek peta
peta = folium.Map(location=TitikAwal, zoom_start=12)

# menampilkan marker titik awal dan titik akhir dengan keterangan Node A dan Node L
folium.Marker(location=TitikAwal, tooltip="Node A",icon=folium.Icon(color='green')).add_to(peta)
folium.Marker(location=TitikAkhir, tooltip="Node L",icon=folium.Icon(color='red')).add_to(peta)

# menampilkan garis penghubung dari titik awal ke titik akhir dengan warna hitam
folium.PolyLine(locations=[TitikAwal, TitikAkhir], color='black').add_to(peta)

# tampilkan peta
peta


### Menampilkan Keterangan Sesuai Map di Atas 
---

| Node | Nama Jalan |
| -------- | -------- |
| A | Jalan Maribaya|
| L | Jalan Dipatiukur (UNIKOM)|

---

### Menampilkan Semua Titik Koordinat Lalu Hubungkan dengan Garis

In [10]:
import folium

# definisikan koordinat node
A = -6.81458,107.62306
B = -6.81495,107.61428
C = -6.81752,107.62242
D = -6.82171,107.63139
E = -6.82271,107.62295
F = -6.82675,107.62552
G = -6.82872,107.63081
H = -6.84915,107.61691
I = -6.86230,107.59622
J = -6.86638,107.62102
K = -6.88335,107.60498
L = -6.88701,107.61552

# buat objek peta
peta = folium.Map(location=A, zoom_start=12)

# tambahkan marker untuk setiap node
nodes = [A, B, C, D, E, F, G, H, I, J, K, L]
node_names = ['Node A', 
              'Node B', 
              'Node C', 
              'Node D', 
              'Node E', 
              'Node F', 
              'Node G', 
              'Node H', 
              'Node I', 
              'Node J', 
              'Node K', 
              'Node L']

for i in range(len(nodes)):
    if i == 0: # node A
        folium.Marker(location=nodes[i], tooltip=node_names[i], icon=folium.Icon(color='green')).add_to(peta)
    elif i == len(nodes)-1: # node L
        folium.Marker(location=nodes[i], tooltip=node_names[i], icon=folium.Icon(color='red')).add_to(peta)
    elif i > 0 and i < 11: # node B-K
        folium.Marker(location=nodes[i], tooltip=node_names[i], icon=folium.Icon(color='blue')).add_to(peta)
    else:
        folium.Marker(location=nodes[i], tooltip=node_names[i]).add_to(peta)

# hubungkan setiap node dengan jalur
# definisikan urutan node dalam jalur
jalur = [A, C, B, I, K, L, J, G, D, C, E, F, H, K, I, E, F, G, J, H]

# tambahkan jalur ke peta
folium.PolyLine(locations=jalur, color='black').add_to(peta)

# tampilkan peta
peta

### Menampilkan Keterangan Sesuai Map di Atas 
---

| Node | Nama Jalan |
| -------- | -------- |
| A | Jalan Maribaya|
| B | Jalan Cijeruk|
| C | Jalan Grand Hotel|
| D | Jalan Kayu Ambon|
| E | Jalan Cijeruk|
| F | Jalan Punclut|
| G | Jalan Pager Wangi|
| H | Jalan Citra Green|
| I | Jalan Ir. H. Djuanda|
| J | Jalan Dr. Setiabudi|
| K | Jalan Ciumbuleuit|
| L | Jalan Dipatiukur (UNIKOM)|

---

## Menentukan Jarak Masing-masing Node

### Jarak A ke C

In [11]:
from geopy.distance import geodesic

# koordinat
A = (-6.81458, 107.62306)
C = (-6.81752, 107.62242)

# menjumlahkan
distance = geodesic(A, C).km

print(f"Jarak antara A dan C adalah {distance:.2f} km.")

Jarak antara A dan C adalah 0.33 km.


### Jarak C ke B

In [12]:
from geopy.distance import geodesic

# # koordinat
C = (-6.81752, 107.62242)
B = -6.81495,107.61428

# menjumlahkan
distance = geodesic(C, B).km

print(f"Jarak antara C dan B adalah {distance:.2f} km.")

Jarak antara C dan B adalah 0.94 km.


### Jarak B ke I

In [13]:
from geopy.distance import geodesic

# koordinat
B = -6.81495,107.61428
I = -6.86230,107.59622

# menjumlahkan
distance = geodesic(B, I).km

print(f"Jarak antara B dan I adalah {distance:.2f} km.")

Jarak antara B dan I adalah 5.60 km.


### Jarak I ke K

In [14]:
from geopy.distance import geodesic

# koordinat
I = -6.86230,107.59622
K = -6.88335,107.60498

# menjumlahkan
distance = geodesic(I, K).km

print(f"Jarak antara I dan K adalah {distance:.2f} km.")

Jarak antara I dan K adalah 2.52 km.


### Jarak K ke L

In [15]:
from geopy.distance import geodesic

# koordinat 
K = -6.88335,107.60498
L = -6.88704,107.61508

# menjumlahkan
distance = geodesic(K, L).km

print(f"Jarak antara K dan L adalah {distance:.2f} km.")

Jarak antara K dan L adalah 1.19 km.


### Jarak C ke E

In [16]:
from geopy.distance import geodesic

# koordinat
C = -6.81752,107.62242
E = -6.82271,107.62295

# menjumlahkan
distance = geodesic(C, E).km

print(f"Jarak antara C dan E adalah {distance:.2f} km.")

Jarak antara C dan E adalah 0.58 km.


### Jarak E ke I

In [17]:
from geopy.distance import geodesic

# koordinat
E = -6.82271,107.62295
I = -6.86230,107.59622

# menjumlahkan
distance = geodesic(E, I).km

print(f"Jarak antara E dan I adalah {distance:.2f} km.")

Jarak antara E dan I adalah 5.28 km.


### Jarak E ke F

In [18]:
from geopy.distance import geodesic

# koordinat
E = -6.82271,107.62295
F = -6.82675,107.62552

# menjumlahkan
distance = geodesic(E, F).km

print(f"Jarak antara E dan F adalah {distance:.2f} km.")

Jarak antara E dan F adalah 0.53 km.


### Jarak F ke G

In [19]:
from geopy.distance import geodesic

# koordinat
F = -6.82675,107.62552
G = -6.82872,107.63081

# menjumlahkan
distance = geodesic(F, G).km

print(f"Jarak antara F dan G adalah {distance:.2f} km.")

Jarak antara F dan G adalah 0.62 km.


### Jarak F ke H

In [20]:
from geopy.distance import geodesic

# koordinat
F = -6.82675,107.62552
H = -6.84915,107.61691

# menjumlahkan
distance = geodesic(F, H).km

print(f"Jarak antara F dan H adalah {distance:.2f} km.")

Jarak antara F dan H adalah 2.65 km.


### Jarak H ke K

In [21]:
from geopy.distance import geodesic

# koordinat
H = -6.84915,107.61691
K = -6.88335,107.60498

# menjumlahkan
distance = geodesic(H, K).km

print(f"Jarak antara H dan K adalah {distance:.2f} km.")

Jarak antara H dan K adalah 4.01 km.


### Jarak H ke J

In [22]:
from geopy.distance import geodesic

# koordinat
H = -6.84915,107.61691
J = -6.86638,107.62102

# menjumlahkan
distance = geodesic(H, J).km

print(f"Jarak antara H dan J adalah {distance:.2f} km.")

Jarak antara H dan J adalah 1.96 km.


### Jarak C ke D

In [23]:
from geopy.distance import geodesic

# koordinat
C = -6.81752,107.62242
D = -6.82171,107.63139

# menjumlahkan
distance = geodesic(C, D).km

print(f"Jarak antara C dan D adalah {distance:.2f} km.")

Jarak antara C dan D adalah 1.09 km.


### Jarak D ke G

In [24]:
from geopy.distance import geodesic

# koordinat
D = -6.82171,107.63139
G = -6.82872,107.63081

# menjumlahkan
distance = geodesic(D, G).km

print(f"Jarak antara D ke G adalah {distance:.2f} km.")

Jarak antara D ke G adalah 0.78 km.


### Jarak G ke J

In [25]:
from geopy.distance import geodesic

# koordinat
G = -6.82872,107.63081
J = -6.86638,107.62102

# menjumlahkan
distance = geodesic(G, J).km

print(f"Jarak antara G ke J adalah {distance:.2f} km.")

Jarak antara G ke J adalah 4.30 km.


### Jarak J ke L

In [26]:
from geopy.distance import geodesic

# koordinat
J = -6.86638,107.62102
L = -6.88704,107.61508

# menjumlahkan
distance = geodesic(J, L).km

print(f"Jarak antara J ke L adalah {distance:.2f} km.")

Jarak antara J ke L adalah 2.38 km.


### Hasil Jarak antar node 

---

| Node | Jumlah Jarak | Satuan |
| -------- | -------- | -------- | 
| A-C | 0.33 | km |
| C-B | 0.94 | km |
| B-I | 5.60 | km |
| I-K | 2.52 | km |
| K-L | 1.19 | km |
| C-E | 0.58 | km |
| E-F | 0.53 | km |
| E-I | 5.28 | km |
| F-G | 0.62 | km |
| F-H | 2.65 | km |
| H-J | 1.96 | km |
| H-K | 4.01 | km |
| C-D | 1.09 | km |
| D-G | 0.78 | km |
| G-J | 4.30 | km |
| J-L | 2.38 | km |

---

### Menampilkan Koordinat Node dan Nama Node

In [28]:
import folium
from haversine import haversine

# definisikan koordinat node
A = -6.81458,107.62306
B = -6.81495,107.61428
C = -6.81752,107.62242
D = -6.82171,107.63139
E = -6.82271,107.62295
F = -6.82675,107.62552
G = -6.82872,107.63081
H = -6.84915,107.61691
I = -6.86230,107.59622
J = -6.86638,107.62102
K = -6.88335,107.60498
L = -6.88704,107.61508

# buat kamus untuk memudahkan pencarian koordinat node berdasarkan nama node
nodes = {
    'A': A,
    'B': B,
    'C': C,
    'D': D,
    'E': E,
    'F': F,
    'G': G,
    'H': H,
    'I': I,
    'J': J,
    'K': K,
    'L': L
}

### Menampilkan Jarak Node A-C Beserta Mapnya

In [29]:
# minta input node awal dari pengguna (hanya boleh A)
while True:
    node_a = input("Inputkan nama node awal (A): ")
    if node_a == 'A':
        break
    print("Node awal yang diinputkan salah! Hanya boleh menginputkan node A dengan huruf kapital.")

# minta input node akhir dari pengguna (hanya boleh C)
while True:
    node_c = input("Inputkan nama node akhir (C): ")
    if node_c == 'C':
        break
    print("Node akhir yang diinputkan salah! Hanya boleh menginputkan node C dengan huruf kapital.")

# hitung jarak antara dua node berdasarkan koordinatnya
jarak = haversine(nodes[node_a], nodes[node_c])

# tampilkan jarak
print(f"Jarak Terpendek {node_a}-{node_c} adalah {jarak:.2f} km.")

# Define the map center
map_center = (-6.83333, 107.61667)

# Create a map object
m = folium.Map(location=map_center, zoom_start=12)

# Add markers for each node
for node_name, node_location in nodes.items():
    if node_name == 'A':
        folium.Marker(location=node_location, tooltip=node_name, icon=folium.Icon(color='green')).add_to(m)
    elif node_name == 'L':
        folium.Marker(location=node_location, tooltip=node_name, icon=folium.Icon(color='red')).add_to(m)
    else:
        folium.Marker(location=node_location, tooltip=node_name).add_to(m)
    
# Get the coordinates of the starting and ending nodes
start_location = nodes[node_a]
end_location = nodes[node_c]

# Create a polyline between the starting and ending nodes
folium.PolyLine(locations=[start_location, end_location], color='red').add_to(m)

# Prompt the user to choose whether to display the map or not
tampilkan_map = input("Tampilkan map? (y/n): ")

# Display the map if the user chooses to
if tampilkan_map == 'y':
    display(m)
elif tampilkan_map == 'n':
    print("Map tidak ditampilkan.")
else:
    print("Pilihan tidak valid. Map tidak ditampilkan.")


Inputkan nama node awal (A): A
Inputkan nama node akhir (C): C
Jarak Terpendek A-C adalah 0.33 km.
Tampilkan map? (y/n): y


### Membandingkan Node ACB, Node ACD, dan ACE
### Rute yang Paling Pendek akan Menampilkan Jumlah Jarak dan Mapnya

In [30]:
# minta input node awal dari pengguna (hanya boleh A)
while True:
    node_a = input("Inputkan nama node awal (A): ")
    if node_a == 'A':
        break
    print("Node awal yang diinputkan salah! Hanya boleh menginputkan node A dengan huruf kapital.")

while True:
    node_b = input("Inputkan nama node tambahan (B): ")
    if node_b == 'B':
        break
    print("Node tambahan yang diinputkan salah! Hanya boleh menginputkan node B dengan huruf kapital.")
    
# minta input node akhir dari pengguna (hanya boleh C)
while True:
    node_c = input("Inputkan nama node akhir (C): ")
    if node_c == 'C':
        break
    print("Node akhir yang diinputkan salah! Hanya boleh menginputkan node C dengan huruf kapital.")

while True:
    node_d = input("Inputkan nama node tambahan (D): ")
    if node_d == 'D':
        break
    print("Node tambahan yang diinputkan salah! Hanya boleh menginputkan node D dengan huruf kapital.")

while True:
    node_e = input("Inputkan nama node tambahan (E): ")
    if node_e == 'E':
        break
    print("Node tambahan yang diinputkan salah! Hanya boleh menginputkan node E dengan huruf kapital.")
    
# hitung jarak antara dua node berdasarkan koordinatnya
jarak_AC = haversine(nodes[node_a], nodes[node_c])
jarak_CB = haversine(nodes[node_c], nodes[node_b])
jarak_ACB = jarak_AC + jarak_CB

jarak_AC = haversine(nodes[node_a], nodes[node_c])
jarak_CE = haversine(nodes[node_c], nodes[node_e])
jarak_ACE = jarak_AC + jarak_CE

jarak_AC = haversine(nodes[node_a], nodes[node_c])
jarak_CD = haversine(nodes[node_c], nodes[node_d])
jarak_ACD = jarak_AC + jarak_CD

if jarak_ACB < jarak_ACE and jarak_ACB < jarak_ACD :
    print(f"Jarak terpendek adalah A-C-B sebesar {jarak_ACB:.2f} km.")
elif jarak_ACE < jarak_ACB and jarak_ACE < jarak_ACD:
    print(f"Jarak terpendek adalah A-C-E sebesar {jarak_ACE:.2f} km.")
else:
    print(f"Jarak terpendek adalah A-C-D sebesar {jarak_ACD:.2f} km.")

# Define the map center
map_center = (-6.83333, 107.61667)

# Create a map object
m = folium.Map(location=map_center, zoom_start=12)

# Add markers for each node
for node_name, node_location in nodes.items():
    if node_name == 'A':
        folium.Marker(location=node_location, tooltip=node_name, icon=folium.Icon(color='green')).add_to(m)
    elif node_name == 'L':
        folium.Marker(location=node_location, tooltip=node_name, icon=folium.Icon(color='red')).add_to(m)
    else:
        folium.Marker(location=node_location, tooltip=node_name).add_to(m)
    
# Get the coordinates of the starting and ending nodes
Node_A = nodes[node_a]
Node_C = nodes[node_c]
Node_E = nodes[node_e]

# Create a polyline between the starting and ending nodes
folium.PolyLine(locations=[Node_A, Node_C, Node_E], color='red').add_to(m)

# Prompt the user to choose whether to display the map or not
tampilkan_map = input("Tampilkan map? (y/n): ")

# Display the map if the user chooses to
if tampilkan_map == 'y':
    display(m)
elif tampilkan_map == 'n':
    print("Map tidak ditampilkan.")
else:
    print("Pilihan tidak valid. Map tidak ditampilkan.")

Inputkan nama node awal (A): A
Inputkan nama node tambahan (B): B
Inputkan nama node akhir (C): C
Inputkan nama node tambahan (D): D
Inputkan nama node tambahan (E): E
Jarak terpendek adalah A-C-E sebesar 0.91 km.
Tampilkan map? (y/n): y


### Membandingkan Node A-C-E-F dan A-C-E-I
### Rute yang Paling Pendek akan Menampilkan Jumlah Jarak dan Mapnya

In [31]:
# minta input node awal dari pengguna (hanya boleh A)
while True:
    node_a = input("Inputkan nama node awal (A): ")
    if node_a == 'A':
        break
    print("Node awal yang diinputkan salah! Hanya boleh menginputkan node A dengan huruf kapital.")

# minta input node akhir dari pengguna (hanya boleh C)
while True:
    node_c = input("Inputkan nama node akhir (C): ")
    if node_c == 'C':
        break
    print("Node akhir yang diinputkan salah! Hanya boleh menginputkan node C dengan huruf kapital.")

while True:
    node_e = input("Inputkan nama node tambahan (E): ")
    if node_e == 'E':
        break
    print("Node tambahan yang diinputkan salah! Hanya boleh menginputkan node E dengan huruf kapital.")

while True:
    node_f = input("Inputkan nama node tambahan (F): ")
    if node_f == 'F':
        break
    print("Node tambahan yang diinputkan salah! Hanya boleh menginputkan node F dengan huruf kapital.")

while True:
    node_i = input("Inputkan nama node tambahan (I): ")
    if node_i == 'I':
        break
    print("Node tambahan yang diinputkan salah! Hanya boleh menginputkan node I dengan huruf kapital.")

# hitung jarak antara dua node berdasarkan koordinatnya
jarak_AC = haversine(nodes[node_a], nodes[node_c])
jarak_CE = haversine(nodes[node_c], nodes[node_e])
jarak_EF = haversine(nodes[node_e], nodes[node_f])
jarak_ACEF = jarak_AC + jarak_CE + jarak_EF 

jarak_AC = haversine(nodes[node_a], nodes[node_c])
jarak_CE = haversine(nodes[node_c], nodes[node_e])
jarak_EI = haversine(nodes[node_e], nodes[node_i])
jarak_ACEI = jarak_AC + jarak_CE + jarak_EI 

if jarak_ACEF < jarak_ACEI:
    print(f"Jarak terpendek adalah A-C-E-F sebesar {jarak_ACEF:.2f} km.")
else:
    print(f"Jarak terpendek adalah A-C-E-I sebesar {jarak_ACEI:.2f} km.")

# Define the map center
map_center = (-6.83333, 107.61667)

# Create a map object
m = folium.Map(location=map_center, zoom_start=12)

# Add markers for each node
for node_name, node_location in nodes.items():
    if node_name == 'A':
        folium.Marker(location=node_location, tooltip=node_name, icon=folium.Icon(color='green')).add_to(m)
    elif node_name == 'L':
        folium.Marker(location=node_location, tooltip=node_name, icon=folium.Icon(color='red')).add_to(m)
    else:
        folium.Marker(location=node_location, tooltip=node_name).add_to(m)
    
# Get the coordinates of the starting and ending nodes
Node_A = nodes[node_a]
Node_C = nodes[node_c]
Node_E = nodes[node_e]
Node_F = nodes[node_f]

# Create a polyline between the starting and ending nodes
folium.PolyLine(locations=[Node_A, Node_C, Node_E, Node_F], color='red').add_to(m)

# Prompt the user to choose whether to display the map or not
tampilkan_map = input("Tampilkan map? (y/n): ")

# Display the map if the user chooses to
if tampilkan_map == 'y':
    display(m)
elif tampilkan_map == 'n':
    print("Map tidak ditampilkan.")
else:
    print("Pilihan tidak valid. Map tidak ditampilkan.")

Inputkan nama node awal (A): A
Inputkan nama node akhir (C): C
Inputkan nama node tambahan (E): E
Inputkan nama node tambahan (F): F
Inputkan nama node tambahan (I): I
Jarak terpendek adalah A-C-E-F sebesar 1.45 km.
Tampilkan map? (y/n): y


### Membandingkan Node A-C-E-F-G dan A-C-E-F-H
### Rute yang Paling Pendek akan Menampilkan Jumlah Jarak dan Mapnya

In [32]:
# minta input node awal dari pengguna (hanya boleh A)
while True:
    node_a = input("Inputkan nama node awal (A): ")
    if node_a == 'A':
        break
    print("Node awal yang diinputkan salah! Hanya boleh menginputkan node A dengan huruf kapital.")

# minta input node akhir dari pengguna (hanya boleh C)
while True:
    node_c = input("Inputkan nama node akhir (C): ")
    if node_c == 'C':
        break
    print("Node akhir yang diinputkan salah! Hanya boleh menginputkan node C dengan huruf kapital.")

while True:
    node_e = input("Inputkan nama node tambahan (E): ")
    if node_e == 'E':
        break
    print("Node tambahan yang diinputkan salah! Hanya boleh menginputkan node E dengan huruf kapital.")

while True:
    node_f = input("Inputkan nama node tambahan (F): ")
    if node_f == 'F':
        break
    print("Node tambahan yang diinputkan salah! Hanya boleh menginputkan node F dengan huruf kapital.")

while True:
    node_g = input("Inputkan nama node tambahan (G): ")
    if node_g == 'G':
        break
    print("Node tambahan yang diinputkan salah! Hanya boleh menginputkan node I dengan huruf kapital.")

while True:
    node_h = input("Inputkan nama node tambahan (H): ")
    if node_h == 'H':
        break
    print("Node tambahan yang diinputkan salah! Hanya boleh menginputkan node I dengan huruf kapital.")

# hitung jarak antara dua node berdasarkan koordinatnya
jarak_AC = haversine(nodes[node_a], nodes[node_c])
jarak_CE = haversine(nodes[node_c], nodes[node_e])
jarak_EF = haversine(nodes[node_e], nodes[node_f])
jarak_FG = haversine(nodes[node_f], nodes[node_g])
jarak_ACEFG = jarak_AC + jarak_CE + jarak_EF  + jarak_FG 

jarak_AC = haversine(nodes[node_a], nodes[node_c])
jarak_CE = haversine(nodes[node_c], nodes[node_e])
jarak_EF = haversine(nodes[node_e], nodes[node_f])
jarak_FH = haversine(nodes[node_f], nodes[node_h])
jarak_ACEFH = jarak_AC + jarak_CE + jarak_EF  + jarak_FH 

if jarak_ACEFG < jarak_ACEFH:
    print(f"Jarak terpendek adalah A-C-E-F-G sebesar {jarak_ACEFG:.2f} km.")
else:
    print(f"Jarak terpendek adalah A-C-E-F-H sebesar {jarak_ACEFH:.2f} km.")

# Define the map center
map_center = (-6.83333, 107.61667)

# Create a map object
m = folium.Map(location=map_center, zoom_start=12)

# Add markers for each node
for node_name, node_location in nodes.items():
    if node_name == 'A':
        folium.Marker(location=node_location, tooltip=node_name, icon=folium.Icon(color='green')).add_to(m)
    elif node_name == 'L':
        folium.Marker(location=node_location, tooltip=node_name, icon=folium.Icon(color='red')).add_to(m)
    else:
        folium.Marker(location=node_location, tooltip=node_name).add_to(m)
    
# Get the coordinates of the starting and ending nodes
Node_A = nodes[node_a]
Node_C = nodes[node_c]
Node_E = nodes[node_e]
Node_F = nodes[node_f]
Node_G = nodes[node_g]

# Create a polyline between the starting and ending nodes
folium.PolyLine(locations=[Node_A, Node_C, Node_E, Node_F, Node_G], color='red').add_to(m)

# Prompt the user to choose whether to display the map or not
tampilkan_map = input("Tampilkan map? (y/n): ")

# Display the map if the user chooses to
if tampilkan_map == 'y':
    display(m)
elif tampilkan_map == 'n':
    print("Map tidak ditampilkan.")
else:
    print("Pilihan tidak valid. Map tidak ditampilkan.")

Inputkan nama node awal (A): A
Inputkan nama node akhir (C): C
Inputkan nama node tambahan (E): E
Inputkan nama node tambahan (F): F
Inputkan nama node tambahan (G): G
Inputkan nama node tambahan (H): H
Jarak terpendek adalah A-C-E-F-G sebesar 2.07 km.
Tampilkan map? (y/n): y


### Menampilkan Jarak A-C-E-F-G-J dan Menampilkan Mapnya

In [34]:
# minta input node awal dari pengguna (hanya boleh A)
while True:
    node_a = input("Inputkan nama node awal (A): ")
    if node_a == 'A':
        break
    print("Node awal yang diinputkan salah! Hanya boleh menginputkan node A dengan huruf kapital.")

# minta input node akhir dari pengguna (hanya boleh C)
while True:
    node_c = input("Inputkan nama node akhir (C): ")
    if node_c == 'C':
        break
    print("Node akhir yang diinputkan salah! Hanya boleh menginputkan node C dengan huruf kapital.")

while True:
    node_e = input("Inputkan nama node tambahan (E): ")
    if node_e == 'E':
        break
    print("Node tambahan yang diinputkan salah! Hanya boleh menginputkan node E dengan huruf kapital.")

while True:
    node_f = input("Inputkan nama node tambahan (F): ")
    if node_f == 'F':
        break
    print("Node tambahan yang diinputkan salah! Hanya boleh menginputkan node F dengan huruf kapital.")

while True:
    node_g = input("Inputkan nama node tambahan (G): ")
    if node_g == 'G':
        break
    print("Node tambahan yang diinputkan salah! Hanya boleh menginputkan node G dengan huruf kapital.")

while True:
    node_j = input("Inputkan nama node tambahan (J): ")
    if node_j == 'J':
        break
    print("Node tambahan yang diinputkan salah! Hanya boleh menginputkan node J dengan huruf kapital.")

# hitung jarak antara dua node berdasarkan koordinatnya
jarak_AC = haversine(nodes[node_a], nodes[node_c])
jarak_CE = haversine(nodes[node_c], nodes[node_e])
jarak_EF = haversine(nodes[node_e], nodes[node_f])
jarak_FG = haversine(nodes[node_f], nodes[node_g])
jarak_GJ = haversine(nodes[node_g], nodes[node_j])
jarak_ACEFGJ = jarak_AC + jarak_CE + jarak_EF  + jarak_FG + jarak_GJ 

print(f"Jarak terpendek adalah A-C-E-F-G-J sebesar {jarak_ACEFGJ:.2f} km.")

# Define the map center
map_center = (-6.83333, 107.61667)

# Create a map object
m = folium.Map(location=map_center, zoom_start=12)

# Add markers for each node
for node_name, node_location in nodes.items():
    if node_name == 'A':
        folium.Marker(location=node_location, tooltip=node_name, icon=folium.Icon(color='green')).add_to(m)
    elif node_name == 'L':
        folium.Marker(location=node_location, tooltip=node_name, icon=folium.Icon(color='red')).add_to(m)
    else:
        folium.Marker(location=node_location, tooltip=node_name).add_to(m)
    
# Get the coordinates of the starting and ending nodes
Node_A = nodes[node_a]
Node_C = nodes[node_c]
Node_E = nodes[node_e]
Node_F = nodes[node_f]
Node_G = nodes[node_g]
Node_J = nodes[node_j]

# Create a polyline between the starting and ending nodes
folium.PolyLine(locations=[Node_A, Node_C, Node_E, Node_F, Node_G, Node_J], color='red').add_to(m)

# Prompt the user to choose whether to display the map or not
tampilkan_map = input("Tampilkan map? (y/n): ")

# Display the map if the user chooses to
if tampilkan_map == 'y':
    display(m)
elif tampilkan_map == 'n':
    print("Map tidak ditampilkan.")
else:
    print("Pilihan tidak valid. Map tidak ditampilkan.")

Inputkan nama node awal (A): A
Inputkan nama node akhir (C): C
Inputkan nama node tambahan (E): E
Inputkan nama node tambahan (F): F
Inputkan nama node tambahan (G): G
Inputkan nama node tambahan (J): J
Jarak terpendek adalah A-C-E-F-G-J sebesar 6.39 km.
Tampilkan map? (y/n): n
Map tidak ditampilkan.


### Menampilkan Jarak A-C-E-F-G-J-L dan Menampilkan Mapnya

In [35]:
# minta input node awal dari pengguna (hanya boleh A)
while True:
    node_a = input("Inputkan nama node awal (A): ")
    if node_a == 'A':
        break
    print("Node awal yang diinputkan salah! Hanya boleh menginputkan node A dengan huruf kapital.")

# minta input node akhir dari pengguna (hanya boleh C)
while True:
    node_c = input("Inputkan nama node akhir (C): ")
    if node_c == 'C':
        break
    print("Node akhir yang diinputkan salah! Hanya boleh menginputkan node C dengan huruf kapital.")

while True:
    node_e = input("Inputkan nama node tambahan (E): ")
    if node_e == 'E':
        break
    print("Node tambahan yang diinputkan salah! Hanya boleh menginputkan node E dengan huruf kapital.")

while True:
    node_f = input("Inputkan nama node tambahan (F): ")
    if node_f == 'F':
        break
    print("Node tambahan yang diinputkan salah! Hanya boleh menginputkan node F dengan huruf kapital.")

while True:
    node_g = input("Inputkan nama node tambahan (G): ")
    if node_g == 'G':
        break
    print("Node tambahan yang diinputkan salah! Hanya boleh menginputkan node G dengan huruf kapital.")

while True:
    node_j = input("Inputkan nama node tambahan (J): ")
    if node_j == 'J':
        break
    print("Node tambahan yang diinputkan salah! Hanya boleh menginputkan node J dengan huruf kapital.")

while True:
    node_l = input("Inputkan nama node tambahan (L): ")
    if node_l == 'L':
        break
    print("Node tambahan yang diinputkan salah! Hanya boleh menginputkan node L dengan huruf kapital.")

# hitung jarak antara dua node berdasarkan koordinatnya
jarak_AC = haversine(nodes[node_a], nodes[node_c])
jarak_CE = haversine(nodes[node_c], nodes[node_e])
jarak_EF = haversine(nodes[node_e], nodes[node_f])
jarak_FG = haversine(nodes[node_f], nodes[node_g])
jarak_GJ = haversine(nodes[node_g], nodes[node_j])
jarak_JL = haversine(nodes[node_j], nodes[node_l])
jarak_ACEFGJL = jarak_AC + jarak_CE + jarak_EF  + jarak_FG + jarak_GJ + jarak_JL 

print(f"Jarak terpendek adalah A-C-E-F-G-J sebesar {jarak_ACEFGJL:.2f} km.")

# Define the map center
map_center = (-6.83333, 107.61667)

# Create a map object
m = folium.Map(location=map_center, zoom_start=12)

# Add markers for each node
for node_name, node_location in nodes.items():
    if node_name == 'A':
        folium.Marker(location=node_location, tooltip=node_name, icon=folium.Icon(color='green')).add_to(m)
    elif node_name == 'L':
        folium.Marker(location=node_location, tooltip=node_name, icon=folium.Icon(color='red')).add_to(m)
    else:
        folium.Marker(location=node_location, tooltip=node_name).add_to(m)
    
# Get the coordinates of the starting and ending nodes
Node_A = nodes[node_a]
Node_C = nodes[node_c]
Node_E = nodes[node_e]
Node_F = nodes[node_f]
Node_G = nodes[node_g]
Node_J = nodes[node_j]
Node_L = nodes[node_l]

# Create a polyline between the starting and ending nodes
folium.PolyLine(locations=[Node_A, Node_C, Node_E, Node_F, Node_G, Node_J, Node_L], color='red').add_to(m)

# Prompt the user to choose whether to display the map or not
tampilkan_map = input("Tampilkan map? (y/n): ")

# Display the map if the user chooses to
if tampilkan_map == 'y':
    display(m)
elif tampilkan_map == 'n':
    print("Map tidak ditampilkan.")
else:
    print("Pilihan tidak valid. Map tidak ditampilkan.")

Inputkan nama node awal (A): A
Inputkan nama node akhir (C): C
Inputkan nama node tambahan (E): E
Inputkan nama node tambahan (F): F
Inputkan nama node tambahan (G): G
Inputkan nama node tambahan (J): J
Inputkan nama node tambahan (L): L
Jarak terpendek adalah A-C-E-F-G-J sebesar 8.78 km.
Tampilkan map? (y/n): y


### Kesimpulan
Rute Terpendek Dari Rumah Acep (Jalan Maribaya) ke Kampus UNIKOM (Dipatiukur) adalah A-C-E-F-G-J-L dengan Total Jarak 8.78 km

Rutenya :
1. A = Jalan Maribaya (Rumah Acep)
2. C = Jalan Grand Hotel
3. E = Jalan Cijeruk
4. F = Jalan Punclut
5. G = Jalan Pager Wangi
6. J = Jalan Dr. Setiabudi
7. L = Jalan Dipatiukur (UNIKOM)